source/uri.c

```diff
@@ -350,7 +350,21 @@ static void s_parse_authority(struct uri_parser *parser, struct aws_byte_cursor
                 parser->uri->user = userinfo_parse_csr;
             }
         }
-        uint8_t *port_delim = memchr(authority_parse_csr.ptr, ':', authority_parse_csr.len);
+
+        /* RFC-3986 section 3.2: host identified by IPv6 literal address is
+         * enclosed within square brackets. We must ignore any colons within
+         * IPv6 literals and only search for port delimiter after closing bracket.*/
+        uint8_t *port_search_start = authority_parse_csr.ptr;
+        if (authority_parse_csr.ptr[0] == '[') {
+            port_search_start = memchr(authority_parse_csr.ptr, ']', authority_parse_csr.len);
+            if (!port_search_start) {
+                parser->state = ERROR;
+                aws_raise_error(AWS_ERROR_MALFORMED_INPUT_STRING);
+                return;
+            }
+        }
+
+        uint8_t *port_delim = memchr(port_search_start, ':', authority_parse_csr.len);
 
         if (!port_delim) {
             parser->uri->port = 0;
```

Comment:
commenting on code below this...

if the URI has brackets in it, should the `parser->uri->host_name` also have the brackets in it? Or are we supposed to strip them out when users query later what the host_name is?

Reply:
my feeling on this is that we should preserve ipv6 encoding in host_name. undoing uri should be a separate concern, but lets discuss

Reply:
looking at how other people implement it. its pretty common to keep brackets on authority, but strip them on host_name, thoughts?

Reply:
I want to do whatever the spec says

Reply:
[Spec says](https://www.rfc-editor.org/rfc/rfc3986#section-3.2.2)
```
   authority     = [ userinfo "@" ] host [ ":" port ]
   userinfo      = *( unreserved / pct-encoded / sub-delims / ":" )
   host          = IP-literal / IPv4address / reg-name
   port          = *DIGIT

   IP-literal    = "[" ( IPv6address / IPvFuture  ) "]"
```

Reply:
and we're doing that!
yay